src/Digdir.Domain.Dialogporten.WebApi/Common/Authentication/TokenIssuerCache.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+using Microsoft.IdentityModel.Protocols;
+using System.Collections.ObjectModel;
+
+namespace Digdir.Domain.Dialogporten.WebApi.Common.Authentication;
+
+public interface ITokenIssuerCache
+{
+    public Task<string?> GetIssuerForScheme(string schemeName);
+}
+
+public class TokenIssuerCache : ITokenIssuerCache, IDisposable
+{
+    private readonly Dictionary<string, string> _issuerMappings = new();
+    private readonly SemaphoreSlim _initializationSemaphore = new(1, 1);
+    private bool _initialized;
+    private readonly IEnumerable<JwtBearerTokenSchemasOptions>? _jwtTokenSchemas;
+
+    public TokenIssuerCache(IConfiguration configuration)
+    {
+        _jwtTokenSchemas = configuration
+            .GetSection(WebApiSettings.SectionName)
+            .Get<WebApiSettings>()
+            ?.Authentication
+            ?.JwtBearerTokenSchemas;
+    }
+
+    public async Task<string?> GetIssuerForScheme(string schemeName)
+    {
+        await EnsureInitializedAsync();
+
+        return _issuerMappings.TryGetValue(schemeName, out var issuer)
+            ? issuer : null;
+    }
+
+    private async Task EnsureInitializedAsync()
+    {
+        if (!_initialized && _jwtTokenSchemas != null)
+        {
+            await _initializationSemaphore.WaitAsync();
+            try
+            {
+                if (!_initialized)
```

Comment:
Veldig bra at du sjekker to ganger! 👍 Dog lov å invertere, og du trenger ikke å gå inn i konteksten av try for å ejecte :grin: